src/cpp/src/text2image/schedulers/ddim.cpp

```diff
@@ -0,0 +1,184 @@
+    using utils::read_json_param;
+
+    read_json_param(data, "num_train_timesteps", num_train_timesteps);
+    read_json_param(data, "beta_start", beta_start);
+    read_json_param(data, "beta_end", beta_end);
+    read_json_param(data, "beta_schedule", beta_schedule);
+    read_json_param(data, "trained_betas", trained_betas);
+    read_json_param(data, "clip_sample", clip_sample);
+    read_json_param(data, "set_alpha_to_one", set_alpha_to_one);
+    read_json_param(data, "steps_offset", steps_offset);
+    read_json_param(data, "prediction_type", prediction_type);
+    read_json_param(data, "thresholding", thresholding);
+    read_json_param(data, "dynamic_thresholding_ratio", dynamic_thresholding_ratio);
+    read_json_param(data, "clip_sample_range", clip_sample_range);
+    read_json_param(data, "sample_max_value", sample_max_value);
+    read_json_param(data, "timestep_spacing", timestep_spacing);
+    read_json_param(data, "rescale_betas_zero_snr", rescale_betas_zero_snr);
+}
+
+DDIMScheduler::DDIMScheduler(const std::string scheduler_config_path) 
+    : DDIMScheduler(Config(scheduler_config_path)) {
+}
+
+DDIMScheduler::DDIMScheduler(const Config& scheduler_config)
+    : m_config(scheduler_config) {
+
+    std::vector<float> alphas, betas;
+
+    using utils::linspace;
+
+    if (!m_config.trained_betas.empty()) {
+        betas = m_config.trained_betas;
+    } else if (m_config.beta_schedule == BetaSchedule::LINEAR) {
+        betas = linspace<float>(m_config.beta_start, m_config.beta_end, m_config.num_train_timesteps);
+    } else if (m_config.beta_schedule == BetaSchedule::SCALED_LINEAR) {
+        float start = std::sqrt(m_config.beta_start);
+        float end = std::sqrt(m_config.beta_end);
+        betas = linspace<float>(start, end, m_config.num_train_timesteps);
+        std::for_each(betas.begin(), betas.end(), [] (float & x) { x *= x; });
+    } else {
+        OPENVINO_THROW("'beta_schedule' must be one of 'LINEAR' or 'SCALED_LINEAR'. Please, add support of other types");
+    }
+
+    // TODO: Rescale for zero SNR
+    // if (m_config.rescale_betas_zero_snr) {betas = rescale_zero_terminal_snr(betas)}
+
+    std::transform(betas.begin(), betas.end(), std::back_inserter(alphas), [] (float b) { return 1.0f - b; });
+
+    for (size_t i = 1; i <= alphas.size(); i++) {
+        float alpha_cumprod =
+            std::accumulate(std::begin(alphas), std::begin(alphas) + i, 1.0, std::multiplies<float>{});
+        m_alphas_cumprod.push_back(alpha_cumprod);
+    }
+
+    m_final_alpha_cumprod = m_config.set_alpha_to_one ? 1 : m_alphas_cumprod[0];
+}
+
+void DDIMScheduler::set_timesteps(size_t num_inference_steps) {
+    m_timesteps.clear();
+
+    if (num_inference_steps > m_config.num_train_timesteps) {
+        OPENVINO_THROW("`num_inference_steps cannot be larger than m_config.num_train_timesteps");
+
+    }
+
+    m_num_inference_steps = num_inference_steps;
+
+    // TODO: add linspace and trailing
```

Comment:
is it planned to fix TODO in future PRs? (or at least we need to add OPENVINO_ASSERT that some modes are not supported
or current version is enough to support other SD 1.x, 2.x models?

Reply:
Could you specify a list of models that should be supported? Tested for SD 2.0 and SD 2.1

Reply:
> is it planned to fix TODO in future PRs? (or at least we need to add OPENVINO_ASSERT that some modes are not supported or current version is enough to support other SD 1.x, 2.x models?

Current version is enough to support listed models (added more outputs to description). These TODOs are more like markers in case we meet a model that doesn't work and need to debug this scheduler.